### PR TITLE
Several functions added for dBZ analysis in EnKF

### PR DESCRIPTION
**Pull request by chunhuazhou**
Several functions added for dBZ analysis in EnKF:
1, Add dBZ ncdiag output for EnKF analysis
2, Enhanced binary dBZ diag output for EnKF analysis 
3, Fix spurious analysis increments when assimilating reflectivity (from OU MAP) 
4, Added EnKF interface for read in dBZ from FV3LAM ensemble forecast (from OU MAP)

Fixed a bug for read in subdomain surface restart files

### CODE REVIEW

File: src/enkf/gridio_fv3reg.f90 [hunk 1]
```diff
@@ -476,6 +492,27 @@ subroutine readgriddata(nanal1,nanal2,vars3d,vars2d,n3d,n2d,levels,ndim,ntimes,
...
+                   do j=1,ny_res
+                     do i=1,nx_res
+                       nn=nn+1
+                       vargrid(nn,levels(dbz_ind-1)+k,nb,ne)=max(workvar3d(i,j,nlevs+1-k),0.0_r_kind)
+                     enddo
+                   enddo
+               enddo
+               do k = levels(dbz_ind-1)+1, levels(dbz_ind)
+                   if (nproc .eq. 0)                                               &
+                      write(6,*) 'READFVregional : qnr ',                           &
```
Comment (Wangy1111) [inline src/enkf/gridio_fv3reg.f90]: @chunhuazhou @hu5970 Line 510: Changing "qnr" to "dbz"?

**Comment by Wangy1111** [comment]
I also checked the remaining changes to EnKF and they are good to me. 
Just a comment on the above summary, it should be "OU MAP" not "OU MAPS". Thank you for integrating our work into this branch.

**Comment by hu5970** [comment]
@chunhuazhou There is a bug in src/gsi/gsi_rfv3io_mod.f90

line 2034: if(allocated(sfc1) .and. allocated(sfc))deallocate (dim_id,sfc,sfc1,dim)
It should be:
      if(allocated(sfc1) .and. allocated(sfc)) deallocate (sfc,sfc1)
      if(allocated(dim)) deallocate (dim)

**Comment by hu5970** [comment]
After bug fix, it can repeated the results of GSI hybrid analysis comparing to current RRFS GSI:
conventional with regional ensemble
conventional with global ensemble
dbz with regional ensemble

File: src/gsi/gsi_dbzOper.F90 [hunk 2]
```diff
@@ -99,8 +103,30 @@ subroutine setup_(self, lunin, mype, is, nobs, init_pass,last_pass)
     character(len=len_isis   ):: isis
     integer(i_kind):: nreal,nchanl,ier,nele
     logical:: diagsave
+    integer(i_kind):: lu_diag
+    character(128):: diag_file
+    character(80):: string
+
+    if(nobs == 0) then
+
+       if(mype == 0) then
```
Comment (daviddowellNOAA) [inline src/gsi/gsi_dbzOper.F90:112]: Lines 112-115 can be deleted, if necessary.  They were for debugging purposes only.

File: src/gsi/gsi_dbzOper.F90 [hunk 3]
```diff
@@ -99,8 +103,30 @@ subroutine setup_(self, lunin, mype, is, nobs, init_pass,last_pass)
     character(len=len_isis   ):: isis
     integer(i_kind):: nreal,nchanl,ier,nele
     logical:: diagsave
+    integer(i_kind):: lu_diag
+    character(128):: diag_file
+    character(80):: string
+
+    if(nobs == 0) then
+
```
Comment (daviddowellNOAA) [inline src/gsi/gsi_dbzOper.F90:111]: Lines 112-115 can be deleted. They were for debugging purposes only.

File: src/gsi/gsi_dbzOper.F90 [hunk 4]
```diff
@@ -99,8 +103,30 @@ subroutine setup_(self, lunin, mype, is, nobs, init_pass,last_pass)
...
+       if(mype == 0) then
+          write(6,*) 'init_pass = ', init_pass
+          write(6,*) 'l_use_dbz_directDA = ', l_use_dbz_directDA
+       endif
 
-    if(nobs == 0) return
```
Comment (daviddowellNOAA) [inline src/gsi/gsi_dbzOper.F90:103]: Lines 117-125 are for making sure that diag files start with the date.  This solution is messy, but it works.

**Review by daviddowellNOAA — APPROVED** [review]
And I have more good news to report.  After the minor bug fix mentioned by Ming, my tests of EnKF for RRFS Ensemble assimilation of conventional and radar-reflectivity observations exactly reproduced previous results for my test case.  Thanks Yongming, Chunhua, and Ming for all of your recent contributions!

File: src/enkf/gridio_fv3reg.f90 (hunk 1 above)
Comment (chunhuazhou) [reply src/enkf/gridio_fv3reg.f90]: Done

**Comment by chunhuazhou** [comment]
> 

Hi Yongming, can you be more specific about the summary that should be "OU MAP" not "OU MAPS",  which line in which file? I don't see this in the changes. Thanks!

**Pull request closed without merging**